src/COLAMD.jl

```diff
@@ -0,0 +1,131 @@
+                             COLAMD_ERROR_A_not_present => "rowval is a null pointer",
+                             COLAMD_ERROR_p_not_present => "colptr is a null pointer",
+                             COLAMD_ERROR_nrow_negative => "nrow is negative",
+                             COLAMD_ERROR_ncol_negative => "nccol is negative",
+                             COLAMD_ERROR_nnz_negative => "number of nonzeros in matrix is negative",
+                             COLAMD_ERROR_p0_nonzero => "p[0] is nonzero",
+                             COLAMD_ERROR_A_too_small => "workspace is too small",
+                             COLAMD_ERROR_col_length_negative => "a column has a negative number of entries",
+                             COLAMD_ERROR_row_index_out_of_bounds => "a row index is out of bounds",
+                             COLAMD_ERROR_out_of_memory => "out of memory",
+                             COLAMD_ERROR_internal_error => "internal error",
+                            )                         
+
+mutable struct Colamd{T<:Union{Cint,_Clong}} 
+  knobs :: Vector{Cdouble}
+  stats :: Vector{T}
+
+  function Colamd{T}() where T <: Union{Cint,_Clong}
+    knobs = zeros(Cdouble, COLAMD_KNOBS)
+    stats = zeros(T, COLAMD_STATS)
+    ccall(_colamd_set_defaults, Nothing, (Ptr{Cdouble},), knobs)
+    return new(knobs, stats)
+  end
+end
+
+function show(io :: IO, meta :: Colamd)
+  s  = "dense row parameter: $(meta.knobs[COLAMD_DENSE_ROW])\n"
+  s *= "dense col parameter: $(meta.knobs[COLAMD_DENSE_COL])\n"
+  s *= "aggressive absorption: $(meta.knobs[COLAMD_AGGRESSIVE])\n"
+  s *= "memory defragmentation: $(meta.stats[COLAMD_DEFRAG_COUNT])\n"
+  s *= "status: $(colamd_statuses[meta.stats[COLAMD_STATUS]])\n"
+  print(io, s)
+end
+
+@inline print(io :: IO, meta :: Colamd) = show(io, meta)
+
+for (fn, typ) in ((:_colamd_recommended, Cint), (:_colamd_l_recommended, _Clong))
+  @eval begin
+    function colamd_recommended(nnz::$typ, nrow::$typ, ncol::$typ)
+      len = ccall($fn, $typ,
+                  ($typ, $typ, $typ),
+                   nnz , nrow, ncol)
+      return len
+    end
+  end
+end
+
+for (orderfn, typ) in ((:_colamd, Cint), (:_colamd_l, _Clong))
+
+  @eval begin
+
+    function colamd(A::SparseMatrixCSC{F,$typ}, meta::Colamd{$typ}) where F
+      nrow, ncol = size(A)
+      nnz = A.colptr[end] - 1
+      p = A.colptr .- $typ(1)  # 0-based indexing
+      len = colamd_recommended($typ(nnz), $typ(nrow), $typ(ncol))
+      workspace = zeros($typ, len)
+      workspace[1:length(A.rowval)] .= A.rowval .- $typ(1)
+      valid = ccall($orderfn, $typ,
+                    ($typ, $typ, $typ, $Ptr{$typ}, Ptr{$typ}, Ptr{Cdouble}, Ptr{$typ}),
+                     nrow, ncol, len , workspace , p        , meta.knobs  , meta.stats)
+      Bool(valid) || throw("colamd status: $(colamd_statuses[meta.stats[COLAMD_STATUS]])")
```

Comment:
If this fails because the workspace is too small, is there a guideline to increase its size and call `_colamd` again?

Reply:
It will never happen because `Alen` is computed with `colamd_recommended`. I interfaced `colmand_recommended` for this reason.

Reply:
Ok. I thought that was just an estimate.